iothub_client/samples/iothub_ll_c2d_sample/iothub_ll_c2d_sample.c

```diff
@@ -43,13 +43,53 @@
 #include "certs.h"
 #endif // SET_TRUSTED_CERT_IN_SAMPLES
 
+// Uncomment this define to use Asynchronous ACK of Cloud-to-Device messages.
+// #define USE_C2D_ASYNC_ACK
+
+#ifdef USE_C2D_ASYNC_ACK
+#include "azure_c_shared_utility/singlylinkedlist.h"
+#endif
+
 /* Paste in the your iothub connection string  */
 static const char* connectionString = "[device connection string]";
 
 #define MESSAGE_COUNT        3
 static bool g_continueRunning = true;
 static size_t g_message_recv_count = 0;
 
+#ifdef USE_C2D_ASYNC_ACK
+static SINGLYLINKEDLIST_HANDLE g_cloudMessages;
+
+static bool ack_and_remove_message(const void* item, const void* match_context, bool* continue_processing)
+{
+    IOTHUB_DEVICE_CLIENT_LL_HANDLE device_ll_handle = (IOTHUB_DEVICE_CLIENT_LL_HANDLE)match_context;
+    IOTHUB_MESSAGE_HANDLE message = (IOTHUB_MESSAGE_HANDLE)item;
+
+    const char* messageId;
+    if ((messageId = IoTHubMessage_GetMessageId(message)) == NULL)
+    {
+        messageId = "<unavailable>";
+    }
+
+    (void)printf("Sending ACK for cloud message (Message ID: %s)\r\n", messageId);
+
+    if (IoTHubDeviceClient_LL_SendMessageDisposition(device_ll_handle, message, IOTHUBMESSAGE_ACCEPTED) != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("ERROR: IoTHubDeviceClient_LL_SendMessageDisposition..........FAILED!\r\n");
+    }
+
+    *continue_processing = true;
+
+    return true;
+}
+
+static void acknowledge_cloud_messages(IOTHUB_DEVICE_CLIENT_LL_HANDLE device_ll_handle)
```

Comment:
Done. I added a more expanded explanation:

// `ack_and_remove_message` is a function that is executed by `singlylinkedlist_remove_if` for each list element.
// In this implementation it is used to send a delayed acknowledgement of receipt to Azure IoT Hub for each message.

and right above the call to `IoTHubDeviceClient_LL_SendMessageDisposition` this was added:

    // If using AMQP protocol, this function results in sending a MESSAGE DISPOSITION (ACCEPTED) for the given cloud-to-device message.
    // If using MQTT protocol, a PUBACK is sent for the cloud-to-device message (only) if `IOTHUBMESSAGE_ACCEPTED` is used.
    // If using HTTP protocol no delayed acknowledgement is sent for the cloud-to-device message, as this protocol does not support that.
    // Independent of the protocol used, this function MUST be called by the user application if using delayed acknowledgement of 
    // cloud-to-device messages, as it will free the memory allocated for each of those messages received.